proposals/enum-entries.md

```diff
@@ -0,0 +1,174 @@
+
+### Acknowledgements of the problem
+
+* It is considered a ["design bug"](http://mail.openjdk.java.net/pipermail/compiler-dev/2018-July/012242.html) in Java.
+* [JDK-8073381 need API to get enums values without creating a new arra...](https://bugs.openjdk.java.net/browse/JDK-8073381).
+* Scala 3 [attempted](https://github.com/lampepfl/dotty/issues/6620) to address this issue.
+* Graal [specifically optimizes](https://github.com/oracle/graal/issues/574) errorneous pattern `values()[idx]`.
+* [Memory-Hogging Enum.values() Method](https://dzone.com/articles/memory-hogging-enumvalues-method)
+
+### Examples of performance issues
+
+ * [HttpStatus.resolve allocates HttpStatus.values() once per invocation](https://github.com/spring-projects/spring-framework/issues/26842)
+ * [Kotlin standard library](https://github.com/JetBrains/kotlin/blob/92d200e093c693b3c06e53a39e0b0973b84c7ec5/libraries/stdlib/jvm/src/kotlin/text/CharCategoryJVM.kt#L170)
+ * [kotlinx.serializarion Enum deserializer]( https://github.com/Kotlin/kotlinx.serialization/issues/1372)
+ * [MySQL JDBC  Remove Enum.values() calls to avoid unnecessary array](https://github.com/Microsoft/mssql-jdbc/pull/1065)
+
+## Proposal
+
+The proposal that addresses all of the above is as follows:
+
+* Decommission of `Enum.values()` with the IDE assistance without deprecation.
+* Introduce property `entries: EnumEntries<E>` that returns an unmodifiable list of all enum entries.
+* For already compiled Kotlin enums and Java enums, a special mapping class that contains a pre-allocated list of entries is generated. 
+
+### Decommission of `Enum.values()`
+
+Enums have been existing since the very beginning of Kotlin, and since Java 1.5, released in 2004, meaning that the deprecation
+of `values` using Kotlin's standard [deprecation cycle](https://kotlinlang.org/docs/kotlin-evolution.html#incompatible-changes) 
+will create unnecessary disturbance in the ecosystem and will render a lot of already existing educational materials outdated.
+
+To avoid that, `values()` will be softly decommissioned via IDE assistance:
+
+* `values` will be de-prioritized and eventually removed from IDE auto-completion.
+* Soft warning with an intention to replace call to `values()` with call to `entries` will be introduced.
+* All the corresponding materials, such as Kotlin guides, J2K and tutorials will be adjusted to use `entries` API.
+* Eventually, we are going to re-visit this decision and decide on the further deprecation of `values` API.
+
+### `EnumEntries<E>` type
+
+Effectively, `entries` represents an immutable list of enum entries and can be represented as type `List<E>`.
+To have an ability to further extend Enum's API in a non-intrusive manner that does not involve code-generation
+for each enum, we expose a direct subtype of `List<E>` named `EnumEntries`.
+
+All potential extensions, such as `valueOfOrNull(String)`, `hasValue(String)` can be implemented on the standard library
```

Comment:
Thanks for this great proposal! Could we consider adding `valueOfOrNull(String)` as part/collateral of this instead of a followup? It would be great to have a standard centralized function from the start on the SDK to avoid each project having to define its own extension function and then at some point later having to replace those.
Rationale: https://youtrack.jetbrains.com/issue/KT-42619/Add-a-static-valueOfOrNull-method-to-Enums-similar-to-valueOf

Reply:
It will definitely not be part of the initial 1.8.0 because we do not know a proper solution for this problem, but I consider this a natural continuation of this KEEP.

We have two options:

* Invest more into static interfaces/static extensions (https://youtrack.jetbrains.com/issue/KT-11968/) and then introduce a bunch of convenience extensions on all enums, such as `valueOfOrNull` etc.
* Introduce extensions/members on the proposed `EnumEntries` interface. It is a piece of cake and can be added at any moment after this KEEP, but it has a downside of being a bit too verbose: instead of `MyEnum.valueOfOrNull` one has to write `MyEnum.entries.valueOfOrNull`. Such a solution cures the symptoms, not the disease (inability to extend enums statically), or at least can be considered as such.

Reply:
(The option "generate valueOfOrNull right into enum body like values/entries" is out of the question though as it bloats the code and is neither extensible nor future-proof for such a simple task)

Reply:
It may be possible to get an `Enum` value by index or by name, defining: `operator EnumEntries.get(name: String) : E?'

```kotlin
MyEnum.entries[index]
MyEnum.entries[name]
```

Reply:
This is an interesting suggestion, written it down in the corresponding KT issue

Reply:
@qwwdfsad thanks for the explanation and rationale :)
@fvasco I really like that suggestion as well!